### PR TITLE
[INFRA-2253][WIP] Add docker build environment to package jenkins

### PR DESCRIPTION
**Pull request by olblak**
This PR Introduces the two following elements:

- A docker environment that contains everything needed to create Debian, Redhat and Suse packages and I also up date some build scripts command in order to work from that environment.

- A new Jenkinsfile to automate Jenkins package creation based on the docker environment.

Remark: The default behavior for the Jenkinsfile in this PR is to retrieve the latest weekly version from release.repo.jenkins.io but I defined two environment variable that could be use to change that behavior. Obviously those two parameters could be defined in different ways either with input steps or parametrized jobs.

Those two variables are JENKINS_VERSION and JENKINS_DOWNLOAD_URL
JENKINS_DOWNLOAD_URL can be set to whatever maven repository that contains war files
current default value is release.repo.jenkins.io
JENKINS_VERSION which accepts one of the following value
* Value version that exist on the maven repository specified by JENKINS_DOWNLOAD_URL
* 'stable' which search for the latest version that respect the pattern X.Y.Z like 2.172.3
* 'weekly' which search for the latest version. 

Everything is currently being tested on release.ci.jenkins.io which require a vpn access as explained [here](https://github.com/jenkins-infra/openvpn#howto-get-client-access)

I didn't work on publishing scripts yet as I think they can be reused as is and be improved in a next iteration so right now we can only download packages as job artificats

Todo:

- [x] Packaging Docker Image 
- [x]  Debian Packaging
- [x]  Redhat Packaging
- [x] Suse Packaging
- [x]  Windows Packaging
- DEPRECATED: OSx Packaging

### CODE REVIEW

File: utils/getJenkinsVersion.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if version == 'stable':
+            versions = root.findall('versioning/versions/version')
+
+            for version in versions:
+                if len(version.text.split('.')) == 3:
+                    result = version.text
+
+            print "Latest Stable Jenkins version detected: {}".format(result)
+
+        # Search on maven repositiry latest version of jenkins
```
Comment (daniel-beck) [inline utils/getJenkinsVersion.py]:
```suggestion
        # Search in Maven repository for latest version of Jenkins
```

File: Jenkinsfile [hunk 2]
```diff
@@ -44,61 +44,76 @@ pipeline {
...
         }
         stage('Windows') {
+          agent {
+            label 'jnlp-windows'
+          }
           steps {
-            unstash 'GPG'
-            unstash 'WAR'
-            sh 'echo "Running on windows"'
+            container('jnlp-window'){
```
Comment (daniel-beck) [inline Jenkinsfile]: Is this the actual label? `window`?

File: Jenkinsfile [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+      artifactNumToKeepStr: '5',
+      daysToKeepStr: '30',
+      numToKeepStr: '10'
+    )
+    disableConcurrentBuilds()
+  }
+
+//  ENV JENKINS_VERSION
+//    stable: It represents the latest stable version that satifies version pattern X.Y.Z
+//    weekly: It represents the latest weekly version that satisfies version pattern X.Y
```
[the remark this answers is not in the document]
Comment (olblak) [reply Jenkinsfile]: Basically I am parsing the [maven-metadata](https://repo.jenkins-ci.org/releases/org/jenkins-ci/main/jenkins-war/maven-metadata.xml) where I consider metadata.versioning.release as 'weekly' and for 'stable' I browse metadata.versioning.versions and get the latest version that match X.Y.Z    as define [here|https://github.com/olblak/packaging/blob/master/utils/getJenkinsVersion.py#L29-L8] using maven-metadata.xml makes it tool agnostic

File: deb/build/build.sh [hunk 4]
```diff
@@ -41,10 +41,9 @@ pushd $D
       mv ${f}_ ${ARTIFACTNAME}$(echo $f | cut -b8-)
     done
   popd
-  debuild -us -uc -Zgzip -A
+  debuild -Zgzip -A
```
[the remark this answers is not in the document]
Comment (olblak) [reply deb/build/build.sh]: Maybe there is something that I don't understand but why don't we want to sign the source as we have a signing certificate

File: Jenkinsfile (hunk 2 above)
Comment (olblak) [reply Jenkinsfile]: Just a typo in a placeholder, I don't know yet how I will package windows application :D

File: Jenkinsfile (hunk 3 above)
[the remark this answers is not in the document]
Comment (daniel-beck) [reply Jenkinsfile]: > using maven-metadata.xml makes it tool agnostic

That would mean Maven Index. Not great, but doable in a first release.

The comment should be more explicit about how this works.

**Comment by olblak** [comment]
I split the docker image section in a different git repository located here [jenkins-infra/docker-packaging|https://github.com/jenkins-infra/docker-packaging]

**Comment by olblak** [comment]
I merge this [PR](https://github.com/jenkinsci/packaging/pull/122) into the current one in order to validate windows script changes

**Comment by olblak** [comment]
Oups I did something wrong with the git history

File: bin/indexGenerator.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+            'template': 'index.redhat.html'
+        },
+        'opensuse': {
+            'extension': '.rpm',
+            'template': 'index.opensuse.html'
+        },
+        'war': {
+            'extension': '.war',
+            'template': 'index.war.html'
+        }
```
Comment (oleg-nenashev) [inline bin/indexGenerator.py]: We still need MSI here later, right?

**Comment by oleg-nenashev** [comment]
@olblak  macOS packaging it no longer needed after https://groups.google.com/forum/#!topic/jenkinsci-dev/xc-lDVsr0bQ . I think it removes the remaining blocker

**Comment by oleg-nenashev** [comment]
@olblak #122 is merged. Could you please rebase your pull request to https://github.com/jenkinsci/packaging/tree/INFRA-910-core-release-automation ?

**Comment by oleg-nenashev** [comment]
I changed the target branch here

**Comment by oleg-nenashev** [comment]
Shipping it to the integration branch

**Pull request merged**